specs/codecs/wac/spec.md

```diff
@@ -0,0 +1,192 @@
+---
+title: "WAC Specification"
+navTitle: "Spec"
+---
+
+# Specification: WAC
+
+**Status: Prescriptive - Exploratory **
+
+* [Format](#format)
+* [Implementations](#implementations)
+    * [Go](#go)
+    * [Rust/WASM](#rust/wasm)
+
+WAC supports bidirectional transport to and from the [IPLD Data Model].
+In that way it is both a "complete" IPLD Data Model representation and "fitted" to the IPLD Data Model.
+Terminology taken from [Codecs and Completeness].
+
+It takes some inspiration from [Simple DAG], but differs in ways designed to make it reliably round-trip through the IPLD Data Model without loss of fidelity.
+
+The initial motivating use case for WAC is to be a codec that can reliably transport IPLD Data Model data across WebAssembly boundaries while also not being particularly complex to implement.
+
+## Format
+
+### Spec
+
+This format is a series of typing tokens, constant tokens, and inline value data.
+
+Typing tokens are proceeded with the value data for that type.
+
+Every type value can be parsed knowing only the type and without any outside context
+like the container or positional delimiters.
+
+#### Tokens
+
+TODO: Make constants in code match spec
+
+Note: The existing implementations of WAC in https://github.com/aschmahmann/wasm-ipld (Go + Rust/WASM) have different token numbers that we'll need to synchronize before merge.
+
+The tokens for the codec roughly correspond with the [IPLD Data Model Kinds]
+
+They are:
+	Null   = 0
+	True   = 1
+	False  = 2
+	Int    = 3
+	NInt   = 4
+	Float  = 5
+	String = 6
+	Bytes  = 7
+	List   = 8
+	Map    = 9
+	Link   = 10
+
+| Int | Token |
+|---|---|
+| 0 | TYPE_LINK |
+| 1 | TYPE_INTEGER |
+| 2 | TYPE_NEGATIVE_INTEGER |
+| 3 | TYPE_FLOAT |
+| 4 | TYPE_NEGATIVE_FLOAT |
+| 5 | TYPE_STRING |
+| 6 | TYPE_BINARY |
+| 7 | TYPE_MAP |
+| 8 | TYPE_LIST |
+| 9 | VALUE_NULL |
+| 10 | VALUE_TRUE |
+| 11 | VALUE_FALSE |
+
+#### TYPE_LINK
+
+```
+| 0 | CID |
+```
+
+Note: Simple-DAG put a VARINT_LENGTH before the CID indicating how long it was.
+
+CIDs are self-delimiting so this didn't seem necessary.
+
+#### TYPE_INTEGER
+
+```
+| 1 | VARINT |
+```
+
+#### TYPE_SIGNED_INTEGER
+
+```
+| 2 | VARINT |
+```
+
+#### TYPE_FLOAT
+
+```
+| 3 | MATISSA_LENGTH | VARINT
+```
+
+TODO: Floats (and negative floats) need definition here.
+Making implementations actually bidirectional with respect to the IPLD Data Model seems difficult here.
+
+#### TYPE_NEGATIVE_FLOAT
+
+```
+| 4 | MATISSA_LENGTH | VARINT
+```
+
+#### TYPE_STRING
+
+```
+| 5 | VARINT_LENGTH | STRING
+```
+
+Note: This is essentially the same as Binary, but with a different token flag.
+
+As with the IPLD Data Model itself Strings are basically just Bytes with something to tag them as "string-like" but where UTF-8 is encouraged. See [Strings].
```

Comment:
Perhaps somewhat contentious, but this is effectively what the data model is and since I'm trying to represent the data model accurately this is what we've got.

@vmx @rvagg I suspect this means that if I add more fixtures to https://github.com/ipld/codec-fixtures that the Rust/WASM code will not be able to use the existing Rust library and something else will need to be coded up.

Reply:
I think https://github.com/ipld/ipld/issues/203 is inevitable at some point, it's only the Go-exclusive folks that are torn on this

Reply:
Responded in the thread Volker started https://github.com/ipld/ipld/pull/226#discussion_r909215811